diffkemp/simpll/passes/SeparateCallsToBitcastPass.cpp

```diff
@@ -0,0 +1,138 @@
+                    // Get the bitcasted function.
+                    auto srcFun =
+                            dyn_cast<Function>(BitCast->stripPointerCasts());
+
+                    // Ignore the instruction if the number of
+                    // arguments is lower than the number of parameters.
+                    if (Call->getNumArgOperands()
+                        < srcFun->getFunctionType()->getNumParams()) {
+                        continue;
+                    }
+
+                    // Ignore the instruction if the source function returns
+                    // void and the return value is used.
+                    if (srcFun->getReturnType()->isVoidTy()
+                        && Call->getType()->isVoidTy() && !Call->use_empty()) {
+                        continue;
+                    }
+
+                    // Bitcast all arguments to the types of source function
+                    // parameters. If the number of arguments is higher, use the
+                    // remaining arguments without bitcasting as a vargars is
+                    // present in the call instruction.
+                    std::vector<Value *> newArgs;
+                    auto paramType = srcFun->getFunctionType()->param_begin();
+                    auto endParamType = srcFun->getFunctionType()->param_end();
+                    auto arg = Call->arg_begin();
+                    auto endArg = Call->arg_end();
+
+                    while (paramType != endParamType) {
+                        if ((*arg)->getType() == *paramType) {
+                            newArgs.push_back(*arg);
+                        } else {
+                            // Bitcast the argument so that the types match.
+                            auto newArg = CastInst::Create(
+                                    Instruction::BitCast,
+                                    *arg,
+                                    *paramType,
+                                    "",
+                                    newArgs.empty() ? cast<Instruction>(Call)
+                                                    : cast<Instruction>(
+                                                            newArgs.back()));
+
+                            newArg->setDebugLoc(Call->getDebugLoc());
+                            newArgs.push_back(newArg);
+                        }
+
+                        ++paramType;
+                        ++arg;
+                    }
+
+                    // Add the remaining arguments if there are any.
+                    while (arg != endArg) {
+                        newArgs.push_back(*arg);
+                        ++arg;
+                    }
+
+                    // Create a new call instruction using the
+                    // source function and bitcasted arguments.
+                    auto newCall = CallInst::Create(srcFun, newArgs, "", Call);
+                    newCall->setDebugLoc(Call->getDebugLoc());
```

Comment:
There are a few more properties that can be copied from the original call like the calling convention, tail call mark or attributes (not all of them are currently compared in `DifferentialFunctionComparator`, but this could change in the future).

Reply:
@lenticularis39 could you please point to some code where copying of such attributes is already implemented in DiffKemp? I believe that you have written such a code.

Reply:
It's in `RemoveUnusedReturnValuesPass`.

Reply:
Thanks, I can see it there and will look into it.

Reply:
I believe that this code should be fully reusable (and not only in this scenario). Please, correct me if I am wrong. If this is true, however, I also propose to move this into `Utils.cpp`. https://github.com/viktormalik/diffkemp/blob/f321f58586fe0237e1bd8088800c92661f60d640/diffkemp/simpll/passes/RemoveUnusedReturnValuesPass.cpp#L209-L221

Reply:
That's a good idea - the function can take a `Instruction *`, copy the generic properties and then use `isa` to determine the instruction type for the purpose of copying the more specific ones.